src/python/pants/util/objects.py

```diff
@@ -276,20 +275,28 @@ def _allowed_values(cls):
       return list(cls._singletons.keys())
 
     def __new__(cls, value):
-      """Forward `value` to the .create() factory method.
+      """Create an instance of this enum.
 
-      The .create() factory method is preferred, but forwarding the constructor like this allows us
-      to use the generated enum type both as a type to check against with isinstance() as well as a
-      function to create instances with. This makes it easy to use as a pants option type.
+      :param value: Use this as the enum value. If `value` is an instance of this class, return it,
+                    otherwise it is checked against the enum's allowed values.
       """
-      return cls.create(value)
+      if isinstance(value, cls):
+        return value
+
+      if value not in cls._singletons:
+        raise cls.make_type_error(
+          "Value {!r} for '{}' must be one of: {!r}."
+          .format(value, field_name, cls._allowed_values))
+      return cls._singletons[value]
 
     # TODO: figure out if this will always trigger on primitives like strings, and what situations
     # won't call this __eq__ (and therefore won't raise like we want).
     def __eq__(self, other):
-      """Redefine equality to raise to nudge people to use static pattern matching."""
-      raise self.make_type_error(
-        "enum equality is defined to be an error -- use .resolve_for_enum_variant() instead!")
+      """Redefine equality to avoid accidentally comparing against a non-enum."""
+      if type(self) != type(other):
+        raise self.make_type_error(
+          "enum equality is only defined for instances of the same enum class!")
```

Comment:
Maybe this could note what the passed type was.

Reply:
Done!